File: CS/DXWebApplication3/CustomWebDocumentViewerReportResolver.cs
using DevExpress.XtraPivotGrid.Data;
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.UI.PivotGrid;
using DevExpress.XtraReports.Web.WebDocumentViewer;
using DXWebApplication3.Models;
using DXWebApplication3.Reports;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Web;

namespace DXWebApplication3 {
    public class CustomWebDocumentViewerReportResolver : IWebDocumentViewerReportResolver {
        NWindEntities db = new NWindEntities();

        public XtraReport CreatePivotGridReport() {
            XtraReport1 report = new XtraReport1();
            XRPivotGrid pivotGrid = report.xrPivotGrid1;
            pivotGrid.DataSource = db.ProductReports.ToList();
            var session = HttpContext.Current.Session;

            byte[] collapsedState = (byte[])session["CollapsedState"];
            byte[] layout = (byte[])session["Layout"];

            if (collapsedState != null && layout != null)
                using (MemoryStream collapsedStateMS = new MemoryStream(collapsedState))
                using (MemoryStream layoutMS = new MemoryStream(layout)) {

                    PivotGridData data = ((IPivotGridDataContainer)pivotGrid).Data;
                    new DevExpress.Utils.Serializing.XmlXtraSerializer().DeserializeObject(pivotGrid, layoutMS, "PivotGrid");
                    data.LoadCollapsedStateFromStream(collapsedStateMS);
                }
            return report;
        }

        public XtraReport Resolve(string reportEntry) {
            switch (reportEntry) {
                case "PivotGrid":
                    return CreatePivotGridReport();
            }
            return new XtraReport();
        }
    }
}

File: VB/DXWebApplication3/CustomWebDocumentViewerReportResolver.vb
Imports DevExpress.XtraPivotGrid.Data
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.UI.PivotGrid
Imports DevExpress.XtraReports.Web.WebDocumentViewer
Imports DXWebApplication3.Models
Imports DXWebApplication3.Reports
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Web

Namespace DXWebApplication3
    Public Class CustomWebDocumentViewerReportResolver
        Implements IWebDocumentViewerReportResolver

        Private db As New NWindEntities()

        Public Function CreatePivotGridReport() As XtraReport
            Dim report As New XtraReport1()
            Dim pivotGrid As XRPivotGrid = report.xrPivotGrid1
            pivotGrid.DataSource = db.ProductReports.ToList()
            Dim session = HttpContext.Current.Session

            Dim collapsedState() As Byte = DirectCast(session("CollapsedState"), Byte())
            Dim layout() As Byte = DirectCast(session("Layout"), Byte())

            If collapsedState IsNot Nothing AndAlso layout IsNot Nothing Then
                Using collapsedStateMS As New MemoryStream(collapsedState)
                    Using layoutMS As New MemoryStream(layout)

                        Dim data As PivotGridData = DirectCast(pivotGrid, IPivotGridDataContainer).Data
                        Call (New DevExpress.Utils.Serializing.XmlXtraSerializer()).DeserializeObject(pivotGrid, layoutMS, "PivotGrid")
                        data.LoadCollapsedStateFromStream(collapsedStateMS)
                    End Using
                End Using
            End If
            Return report
        End Function

        Public Function Resolve(ByVal reportEntry As String) As XtraReport Implements IWebDocumentViewerReportResolver.Resolve
            Select Case reportEntry
                Case "PivotGrid"
                    Return CreatePivotGridReport()
            End Select
            Return New XtraReport()
        End Function
    End Class
End Namespace